codeworld-server/src/AndroidExport.hs

```diff
@@ -0,0 +1,83 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+{-
+  Copyright 2017 The CodeWorld Authors. All rights reserved.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+-}
+
+module AndroidExport where
+
+import           Data.Maybe
+import qualified Data.Map as M
+import           System.Process
+import           System.Directory
+import           System.FilePath
+import qualified System.IO.Strict as ST
+import           Text.HTML.TagSoup
+
+import Util
+
+buildAndroid :: BuildMode -> ProgramId -> AppProps -> IO ()
+buildAndroid mode programId appProps = do
+  let appName = fromJust $ M.lookup "appName" appProps
+  initCordovaProject mode programId
+  copySource mode programId
+  setAppName mode programId appName
+  buildApk mode programId
+  return ()
+
+initCordovaProject :: BuildMode -> ProgramId -> IO ()
+initCordovaProject mode programId = do
+  let rootDir = androidRootDir mode
+  checkIfRootExists <- doesDirectoryExist rootDir
+  if not checkIfRootExists
+     then do
+       createDirectory $ androidRootDir mode
+     else return ()
+  let buildDir = androidBuildDir mode programId
```

Comment:
You need to build in a temp dir.  Otherwise you'll risk kicking off simultaneous builds pointed at the same directory, and this can confuse things and cause weird bug reports.  It has happened before.